mllib/src/main/scala/org/apache/spark/ml/optim/WeightedLeastSquares.scala

```diff
@@ -85,73 +101,193 @@ private[ml] class WeightedLeastSquares(
-          "Consider setting fitIntercept=true.")
+        aBarStd(j) = aBar(j) / aStd(j)
+      }
+      j += 1
+    }
+
+    val abBarStd = new Array[Double](numFeatures)
+    j = 0
+    while (j < numFeatures) {
+      if (aStd(j) == 0.0) {
+        abBarStd(j) = 0.0
+      } else {
+        abBarStd(j) = abBar(j) / (aStd(j) * bStd)
+      }
+      j += 1
+    }
+
+    val aaBarStd = new Array[Double](triK)
+    j = 0
+    var kk = 0
+    while (j < numFeatures) {
+      val aStdJ = aStd(j)
+      var i = 0
+      while (i <= j) {
+        val aStdI = aStd(i)
+        if (aStdJ == 0.0 || aStdI == 0.0) {
+          aaBarStd(kk) = 0.0
+        } else {
+          aaBarStd(kk) = aaBarValues(kk) / (aStdI * aStdJ)
+        }
+        kk += 1
+        i += 1
       }
+      j += 1
     }
 
+    val bBarStd = bBar / bStd
+    val bbBarStd = bbBar / (bStd * bStd)
+
+    val effectiveRegParam = regParam / bStd
+    val effectiveL1RegParam = elasticNetParam * effectiveRegParam
+    val effectiveL2RegParam = (1.0 - elasticNetParam) * effectiveRegParam
+
     // add regularization to diagonals
     var i = 0
-    var j = 2
+    j = 2
     while (i < triK) {
-      var lambda = regParam
-      if (standardizeFeatures) {
-        lambda *= aVar(j - 2)
+      var lambda = effectiveL2RegParam
+      if (!standardizeFeatures) {
+        val std = aStd(j - 2)
+        if (std != 0.0) {
+          lambda /= (std * std)
+        } else {
+          lambda = 0.0
+        }
       }
-      if (standardizeLabel && bStd != 0) {
-        lambda /= bStd
+      if (!standardizeLabel) {
+        lambda *= bStd
       }
-      aaValues(i) += lambda
+      aaBarStd(i) += lambda
       i += j
       j += 1
     }
+    val aa = getAtA(aaBarStd, aBarStd)
+    val ab = getAtB(abBarStd, bBarStd)
 
-    val aa = if (fitIntercept) {
-      Array.concat(aaBar.values, aBar.values, Array(1.0))
-    } else {
-      aaBar.values
-    }
-    val ab = if (fitIntercept) {
-      Array.concat(abBar.values, Array(bBar))
+    val solver = if ((solverType == WeightedLeastSquares.Auto && elasticNetParam != 0.0) ||
+      (solverType == WeightedLeastSquares.QuasiNewton)) {
+      val effectiveL1RegFun: Option[(Int) => Double] = if (effectiveL1RegParam != 0.0) {
+        Some((index: Int) => {
+            if (fitIntercept && index == numFeatures) {
+              0.0
+            } else {
+              if (standardizeFeatures) {
+                effectiveL1RegParam
+              } else {
+                if (aStd(index) != 0.0) effectiveL1RegParam / aStd(index) else 0.0
+              }
+            }
+          })
+      } else {
+        None
+      }
+      new QuasiNewtonSolver(fitIntercept, maxIter, tol, effectiveL1RegFun)
     } else {
-      abBar.values
+      new CholeskySolver(fitIntercept)
     }
 
-    val x = CholeskyDecomposition.solve(aa, ab)
+    val solution = solver match {
+      case cholesky: CholeskySolver =>
+        try {
```

Comment:
I don't really like this solution. I'd appreciate feedback on alternatives here.